experimental/InvariantTheory/src/InvariantTheory.jl

```diff
@@ -0,0 +1,523 @@
+        if sym != :SL
+            error("Only implemented for SLm")
+        end
+        fld = base_ring(pring)
+        characteristic(fld) == 0 || error("Characteristic should be 0 for linearly reductive groups")
+        G.field = fld
+        @assert m^2  == ngens(pring)
+        G.group = (sym,m)
+        G.reynolds_operator = reynolds_slm
+        M = matrix(pring, m, m, gens(pring))
+        G.canonical_representation = M
+        G.group_ideal = ideal([det(M) - 1])
+        #base ring of M has to be the same as the representation matrix when that is created later.
+        return G
+    end
+end
+
+function Base.show(io::IO, G::ReductiveGroup)
+    io = pretty(io)
+    if G.group[1] == :SL
+        print(io, "Reductive group ", G.group[1], G.group[2])
+    end
+end
+
+#getter functions
+reductive_group(sym::Symbol, m::Int, R::MPolyRing) = ReductiveGroup(sym,m,R)
+reductive_group(sym::Symbol, m::Int, F::Field) =  ReductiveGroup(sym,m,F)
+group(G::ReductiveGroup) = G.group
+field(G::ReductiveGroup) = G.field
+reynolds_operator(G::ReductiveGroup) = G.reynolds_operator
+group_ideal(G::ReductiveGroup) = G.group_ideal
+canonical_representation(G::ReductiveGroup) = G.canonical_representation
+natural_representation(G::ReductiveGroup) = G.canonical_representation
+
+#####################
+#Setting up Representation objects
+#####################
+#Objects of type ReductiveGroup can be embedded in GLn (for some n) via a representation. This defines the action on a vector space. 
+#We set up an object RepresentationReductiveGroup that carries information about this representation.
+
+mutable struct RepresentationReductiveGroup
+    group::ReductiveGroup
+    rep_mat::MatElem
+    sym_deg::Tuple{Bool, Int}
+    reynolds_v::Function
+    
+    #representation of group G over symmetric degree d
+    function RepresentationReductiveGroup(G::ReductiveGroup, d::Int)
+        R = new()
+        R.group = G
+        R.rep_mat = rep_mat_(G, d)
+        R.sym_deg = (true, d)
+        R.reynolds_v = reynolds_v_slm
+        return R
+    end
+    
+    #matrix M is the representation matrix. does not check M.
+    function RepresentationReductiveGroup(G::ReductiveGroup, M::MatElem)
+        @req base_ring(M) == base_ring(G.group_ideal) "Group ideal and representation matrix must have same parent ring"
+        R = new()
+        R.group = G
+        R.rep_mat = M
+        R.sym_deg = (false, 0)
+        R.reynolds_v = reynolds_v_slm
+        return R
+    end
+end
+
+representation_reductive_group(G::ReductiveGroup, M::MatElem) = RepresentationReductiveGroup(G,M)
+group(R::RepresentationReductiveGroup) = R.group
+
+function representation_on_forms(G::ReductiveGroup, d::Int)
+    @assert G.group[1] == :SL
+        return RepresentationReductiveGroup(G, d)
+end
+
+function representation_reductive_group(G::ReductiveGroup)
+    if G.group[1] == :SL
+        M = canonical_representation(G)
+        return RepresentationReductiveGroup(G,M)
+    end
```

Comment:
This function still sometimes returns `nothing`. Please check for this in all your functions